
# The guide to annotate toxin genes in genomes

## Summary

- [Introduction](#introduction)
- [Toxin annotation](#toxin-annotation)
	- [ToxCodAnGenome](#toxcodangenome)
	- [Checking annotations](#checking-annotations)
- [NonToxin annotation](#nontoxin-annotation)
- [Extra: Plotting toxin loci](#plotting-toxin-loci)

# Introduction
The Guide to annotate toxin genes in genomes is part of [**ToxCodAn-Genome**](https://github.com/pedronachtigall/ToxCodAn-Genome) and designed to walk you through our toxin annotation pipeline.

Before start and walking through the guide, it will be good to have some basic knowledge about bioinformatics.

If you don't have much experience in using command lines, running programs, or bioinformatics in general, then you can follow the ["Basic Bioinformatics"](https://github.com/pedronachtigall/ToxCodAn/tree/master/Guide#basic-bioinformatics) section in the ToxCodAn's guide to venom gland transcriptoms. This section has some general resources and information that may help you to better understand bioinformatics and the rest of this Guide. The Guide is not designed to teach you *everything*, so we highly recommend working through the training resources first!

# Toxin annotation

## ToxCodAnGenome

### Toxin database

We have designed toxin database with curated toxin CDSs for ..., which can be downloaded [here](https://github.com/pedronachtigall/ToxCodAn-Genome/tree/main/database/).

### Custom toxin database

If you and your research group have been extensively working within the venomous taxa of the species being analyzed and have a well curated set of toxin CDSs available, it can be used as a toxin database or integrate some of the toxin databases available.

If using a custom toxin database, ensure to have the toxin family specificied in the header of each sequence

If using the custom toxin database as the toxin database, just indicate it as the database file with the parameter ```-d```. In this case, ensure you have the toxin family annotated in the header of each toxin CDS after an ```_``` ("underscore") symbol (e.g., ```>Toxin1_SVMP```).

If you want to integrate some of the avilable toxin database, just indicate the custom toxin database with the parameter ```-C```. In this case, you will not need to specify the toxin family in the header; however, it will consider as a generic toxin, by adding a ```_TOXIN``` at the end of each sequence (e.g., ```>Toxin1_TOXIN```). It must be better to annotate the toxin family in each sequence header to facilitate downstream analysis.

#### Curated toxin CDS set

#### Venom tissue transcriptome
If you have a venom tissue transcriptomic data available for the species being annotated, you should consider using this data the toxin annotation step to improve the final set generated by ToxCodAn-Genome. In this sense, you can detect the toxin CDSs and annotate them by following your own pipeline, or follow the guide to venom gland transcriptomics available within [ToxCodAn](https://github.com/pedronachtigall/ToxCodAn/tree/master/Guide#the-guide) tool and published in [Briefings in Bioinformatics](https://doi.org/10.1093/bib/bbab095).

If you and you research group does not have a strong background on performing such transcriptome anntotation, we designed two scripts to help on this task that can be run separetely or integrated into the main ToxCodAn-Genome pipeline.

<details>
<summary>Expand "Transcriptome Assembly" Section</summary>

</details>
<br>

<details>
<summary>Expand "Toxin CDS annotation" Section</summary>

</details>
<br>

### Running ToxCodAn-Genome

## Checking annotations

### Checking reliable annotations

### Checking warning annotations

### Checking matched regions with no annotation

# NonToxin annotation

# Plotting toxin loci
